src/pages/billing/SpendReport/SpendReport.ts

```diff
@@ -31,10 +28,75 @@ const OtherMessaging = ({ cost }) => {
   ])
 }
 
-export const SpendReport = ({ tab, billingProject }) => {
-  const [updating, setUpdating] = useState(false)
-  const [projectCost, setProjectCost] = useState(null)
-  const [costPerWorkspace, setCostPerWorkspace] = useState(
+// Interfaces for internal storage of data
+interface WorkspaceCosts {
+  workspaceNames: string[]
+  computeCosts: number[]
+  storageCosts: number[]
+  otherCosts: number[]
+  numWorkspaces: number
+  costFormatter: any
+}
+
+interface ProjectCost {
+  spend: string
+  compute: string
+  storage: string
+  other: string
+}
+// End of interfaces for internal storage of data
+
+// Interfaces for dealing with the server SpendReport JSON response
+interface CategorySpendData {
+  category: string
+  cost: string
+  credits: string
+  currency: string
+}
+
+interface WorkspaceSpendData {
+  cost: string
+  credits: string
+  currency: string
+  googleProjectId: string
+  subAggregation: { aggregationKey: 'Category'; spendData: CategorySpendData[]}
+  workspace: { name: string; namespace: string }
+}
+
+interface AggregatedSpendData {
+  aggregationKey: 'Workspace' | 'Category'
+}
+
+interface AggregatedWorkspaceSpendData extends AggregatedSpendData{
+  aggregationKey: 'Workspace'
```

Comment:
This all look well laid out and orgainized - particularly the use of the discriminator.

Reply:
But for some reason it isn't providing the sort of enforcement that I would expect… I will add some comments for specific questions.